go/libraries/doltcore/doltdb/commit_itr.go

```diff
@@ -0,0 +1,135 @@
+package doltdb
+
+import (
+	"context"
+	"errors"
+	"github.com/liquidata-inc/dolt/go/store/hash"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+// CommitItr is an interface for iterating over a set of unique commits
+type CommitItr interface {
+	// Next returns the hash of the next commit, and a pointer to that commit.  Implementations of Next must handle
+	// making sure the list of commits returned are unique.  When complete Next will return hash.Hash{}, nil, nil
+	Next(ctx context.Context) (hash.Hash, *Commit, error)
+}
+
+type commitItr struct {
+	ddb         *DoltDB
+	rootCommits []*Commit
+	currentRoot int
+
+	added       map[hash.Hash]bool
+	unprocessed []hash.Hash
+	curr        *Commit
+}
+
+// CommitItrForAllBranches returns a CommitItr which will iterate over all commits in all branches in a DoltDB
```

Comment:
Needs more detail about the order that results are returned in. Looks like this is a depth-first ancestor traversal for each top-level ref?

Reply:
I don't think I want to guarantee an traversal order.  Users should just know they are going to get all unique commits.